libclamav/udf.c

```diff
@@ -0,0 +1,838 @@
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
+    if (NULL == buffer) {
+        goto done;
+    }
+    lastOffset = idx;
+
+    test = (UnallocatedSpaceDescriptor *)buffer;
+    if (UNALLOCATED_SPACE_DESCRIPTOR != test->tag.tagId) {
+        goto done;
+    }
+
+    ret = test;
+    idx += VOLUME_DESCRIPTOR_SIZE;
+
+done:
+    *idxp        = idx;
+    *lastOffsetp = lastOffset;
+
+    return ret;
+}
+
+static TerminatingDescriptor *getTerminatingDescriptor(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    uint8_t *buffer             = NULL;
+    TerminatingDescriptor *ret  = NULL;
+    TerminatingDescriptor *test = NULL;
+    size_t idx                  = *idxp;
+    size_t lastOffset           = *lastOffsetp;
+
+    if (skipEmptyDescriptors(ctx, idxp, lastOffsetp)) {
+        goto done;
+    }
+
+    idx        = *idxp;
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
+    if (NULL == buffer) {
+        goto done;
+    }
+    lastOffset = idx;
+
+    test = (TerminatingDescriptor *)buffer;
+    if (TERMINATING_DESCRIPTOR != test->tag.tagId) {
+        goto done;
+    }
+
+    ret = test;
+    idx += VOLUME_DESCRIPTOR_SIZE;
+
+done:
+    *idxp        = idx;
+    *lastOffsetp = lastOffset;
+
+    return ret;
+}
+
+static LogicalVolumeIntegrityDescriptor *getLogicalVolumeIntegrityDescriptor(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    uint8_t *buffer                        = NULL;
+    LogicalVolumeIntegrityDescriptor *ret  = NULL;
+    LogicalVolumeIntegrityDescriptor *test = NULL;
+    size_t idx                             = *idxp;
+    size_t lastOffset                      = *lastOffsetp;
+
+    if (skipEmptyDescriptors(ctx, idxp, lastOffsetp)) {
+        goto done;
+    }
+
+    idx        = *idxp;
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
+    if (NULL == buffer) {
+        goto done;
+    }
+    lastOffset = idx;
+
+    test = (LogicalVolumeIntegrityDescriptor *)buffer;
+    if (LOGICAL_VOLUME_INTEGRITY_DESCRIPTOR != test->tag.tagId) {
+        goto done;
+    }
+
+    ret = test;
+    idx += VOLUME_DESCRIPTOR_SIZE;
+
+done:
+    *idxp        = idx;
+    *lastOffsetp = lastOffset;
+
+    return ret;
+}
+
+static AnchorVolumeDescriptorPointer *getAnchorVolumeDescriptorPointer(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    uint8_t *buffer                     = NULL;
+    AnchorVolumeDescriptorPointer *ret  = NULL;
+    AnchorVolumeDescriptorPointer *test = NULL;
+    size_t idx                          = *idxp;
+    size_t lastOffset                   = *lastOffsetp;
+
+    if (skipEmptyDescriptors(ctx, idxp, lastOffsetp)) {
+        goto done;
+    }
+
+    idx        = *idxp;
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
```

Comment:
I'm a little concerned if this needs to be unneeded. Or maybe use `fmap_need_off_once()` instead